sam/pileup.go

```diff
@@ -437,3 +438,140 @@ func (p *Pile) String() string {
 	return fmt.Sprintf("RefIdx: %d\tPos: %d\tCountF: %v\tCountR: %v\tInsCountF: %v\tInsCountR: %v\tDelCountF: %v\tDelCountR: %v\tNext: %p\tPrev: %p",
 		p.RefIdx, p.Pos, p.CountF, p.CountR, p.InsCountF, p.InsCountR, p.DelCountF, p.DelCountR, p.next, p.prev)
 }
+
+// consensusType represents the possible types of consensus variants in a Pile,
+// which includes Bases, INDELs, or undefined (in the case where the Pile contains no data).
+type consensusType byte
+
+const (
+	Base      consensusType = 0
+	Insertion consensusType = 1
+	Deletion  consensusType = 2
+	Undefined consensusType = 3
+)
+
+// Consensus contains information on the most observed variant in a Pile struct.
+type Consensus struct {
+	Base      dna.Base
+	Insertion []dna.Base
+	Deletion  int
+	Type      consensusType
+}
+
+// PileConsensus returns a Consensus struct from an input Pile struct
+// representing information about the consensus variant observed in that Pile.
+func PileConsensus(p Pile, substitutionsOnly bool) Consensus {
+	// first we check if consensus is a base
+	var max int = p.CountF[dna.A] + p.CountR[dna.A]
+	tiedConsensus := make([]Consensus, 1)
+	tiedConsensus[0] = Consensus{Base: dna.A, Type: Base}
+
+	max, tiedConsensus = getMaxBase(p, max, dna.C, tiedConsensus)
+	max, tiedConsensus = getMaxBase(p, max, dna.G, tiedConsensus)
+	max, tiedConsensus = getMaxBase(p, max, dna.T, tiedConsensus)
+	//question for Dan: main version checks dna.Gap here, is that necessary if we check the deletion field?
```

Comment:
Not necessary for the new functionality, but it would be necessary to check when running with substitutionsOnly if you want to preserve legacy compatibility.